probe/plugins/registry.go

```diff
@@ -0,0 +1,271 @@
+package plugins
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/fsnotify/fsnotify"
+	"github.com/ugorji/go/codec"
+	"golang.org/x/net/context"
+	"golang.org/x/net/context/ctxhttp"
+
+	"github.com/weaveworks/scope/common/fs"
+	"github.com/weaveworks/scope/common/fswatch"
+	"github.com/weaveworks/scope/common/xfer"
+	"github.com/weaveworks/scope/report"
+)
+
+// Exposed for testing
+var (
+	transport = makeUnixRoundTripper
+)
+
+const (
+	pluginTimeout = 500 * time.Millisecond
+	pluginRetry   = 5 * time.Second
+)
+
+// Registry maintains a list of available plugins by name.
+type Registry struct {
+	root              string
+	apiVersion        string
+	handshakeMetadata map[string]string
+	pluginsBySocket   map[string]*Plugin
+	lock              sync.RWMutex
+	watcher           fswatch.Watcher
+	done              chan struct{}
+}
+
+// NewRegistry creates a new registry which watches the given dir root for new
+// plugins, and adds them.
+func NewRegistry(root, apiVersion string, handshakeMetadata map[string]string) (*Registry, error) {
+	watcher, err := fswatch.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	r := &Registry{
+		root:              root,
+		apiVersion:        apiVersion,
+		handshakeMetadata: handshakeMetadata,
+		pluginsBySocket:   map[string]*Plugin{},
+		watcher:           watcher,
+		done:              make(chan struct{}),
+	}
+	if err := r.addPath(context.Background(), r.root); err != nil {
+		r.Close()
+		return nil, err
+	}
+	go r.loop()
+	return r, nil
+}
+
+// add recursively crawls the path provided, adding it to the watcher, and
+// looking for any existing sockets, loading them as plugins.
+func (r *Registry) addPath(ctx context.Context, path string) error {
```

Comment:
I find it confusing that the same function is used for adding the top directory and inner files generically, specially since the `remove` counterpart is only used for sockets. I think that specialized functions would be easier to read.